flink/src/main/java/org/apache/iceberg/flink/source/StreamingMonitorFunction.java

```diff
@@ -0,0 +1,163 @@
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.SnapshotUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is the single (non-parallel) monitoring task which takes a {@link FlinkInputFormat},
+ * it is responsible for:
+ *
+ * <ol>
+ *     <li>Monitoring snapshots of the Iceberg table.</li>
+ *     <li>Creating the {@link FlinkInputSplit splits} corresponding to the incremental files</li>
+ *     <li>Assigning them to downstream tasks for further processing.</li>
+ * </ol>
+ *
+ * <p>The splits to be read are forwarded to the downstream {@link StreamingReaderOperator}
+ * which can have parallelism greater than one.
+ */
+public class StreamingMonitorFunction extends RichSourceFunction<FlinkInputSplit> implements CheckpointedFunction {
+
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingMonitorFunction.class);
+
+  private static final long DUMMY_START_SNAPSHOT_ID = -1;
+
+  private final TableLoader tableLoader;
+  private final ScanContext ctxt;
+
+  private volatile boolean isRunning = true;
+  private transient Object checkpointLock;
+  private transient Table table;
+  private transient ListState<Long> snapshotIdState;
+  private transient long startSnapshotId;
+
+  public StreamingMonitorFunction(TableLoader tableLoader, ScanContext ctxt) {
+    Preconditions.checkArgument(ctxt.snapshotId() == null && ctxt.asOfTimestamp() == null,
+        "The streaming reader does not support using snapshot-id or as-of-timestamp to select the table snapshot.");
+    Preconditions.checkArgument(ctxt.endSnapshotId() == null,
+        "The streaming reader does not support using end snapshot id.");
+    this.tableLoader = tableLoader;
+    this.ctxt = ctxt;
+  }
+
+  @Override
+  public void initializeState(FunctionInitializationContext context) throws Exception {
+    snapshotIdState = context.getOperatorStateStore().getListState(
+        new ListStateDescriptor<>(
+            "snapshot-id-state",
+            LongSerializer.INSTANCE));
+    tableLoader.open();
+    table = tableLoader.loadTable();
+    if (context.isRestored()) {
+      LOG.info("Restoring state for the {}.", getClass().getSimpleName());
+      startSnapshotId = snapshotIdState.get().iterator().next();
+    } else {
+      Long optionStartSnapshot = ctxt.startSnapshotId();
+      if (optionStartSnapshot != null) {
+        if (!SnapshotUtil.ancestorOf(table, table.currentSnapshot().snapshotId(), optionStartSnapshot)) {
+          throw new IllegalStateException("The option start-snapshot-id " + optionStartSnapshot +
+              " is not an ancestor of the current snapshot");
+        }
+
+        startSnapshotId = optionStartSnapshot;
+      } else {
+        startSnapshotId = DUMMY_START_SNAPSHOT_ID;
```

Comment:
Should we introduce a start strategy config?
https://github.com/stevenzwu/iceberg/blob/flip27IcebergSource/flink/src/main/java/org/apache/iceberg/flink/source/enumerator/ContinuousEnumConfig.java#L35

I am also wondering if we should move the continuous monitoring config outside ScanContext?

Reply:
The current read strategy support `LATEST_SNAPSHOT` and `SPECIFIC_START_SNAPSHOT_ID`.  I think it's great to introduce an extra read strategy config outside  ScanContext when we integrate flink source/sink with flip-27, for now we can let it go.